File: server/lib/listHighlights.js
import path from 'path'
import fs from 'fs/promises'
import { fileURLToPath } from 'url'
import { readLabels } from './labelStore.js'
import { readHidden } from './hiddenStore.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const RECORDINGS_DIR = path.join(__dirname, '..', 'recordings')

/**
 * List highlight candidates: every non-normal labeled clip on disk, newest
 * first, each annotated with its current demo-page visibility.
 *
 * With { includeHidden: false } (the public demo), clips the owner has hidden
 * are omitted entirely. With { includeHidden: true } (the owner's manager
 * view), they are returned with `hidden: true` so the UI can show and toggle
 * them. Both routes share this so the definition of a highlight lives in one
 * place.
 */
export async function listHighlights({ includeHidden }) {
  const [labels, hidden] = await Promise.all([
    readLabels().catch(() => ({})),
    readHidden().catch(() => new Set()),
  ])
  const files = (await fs.readdir(RECORDINGS_DIR)).filter(f => f.endsWith('.webm'))

  const items = await Promise.all(
    files
      .filter(f => labels[f] && labels[f] !== 'normal')
      .filter(f => includeHidden || !hidden.has(f))
      .map(async (filename) => {
        const stat = await fs.stat(path.join(RECORDINGS_DIR, filename))
        return {
          filename,
          label: labels[filename],
          createdAt: stat.mtime.toISOString(),
          size: stat.size,
          hidden: hidden.has(filename),
        }
      })
  )
  items.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt))
  return items
}

File: server_py/app/lib/list_highlights.py
"""Port of server/lib/listHighlights.js.

List highlight candidates: every non-normal labeled clip on disk, newest
first, each annotated with its current demo-page visibility.
"""
import asyncio

from app.config import RECORDINGS_DIR
from app.lib.hidden_store import read_hidden
from app.lib.iso_time import to_iso_millis
from app.lib.label_store import read_labels


async def list_highlights(include_hidden: bool) -> list[dict]:
    try:
        labels = await read_labels()
    except Exception:
        labels = {}
    try:
        hidden = await read_hidden()
    except Exception:
        hidden = set()

    files = [f.name for f in await asyncio.to_thread(lambda: list(RECORDINGS_DIR.iterdir())) if f.name.endswith(".webm")]

    items = []
    for filename in files:
        label = labels.get(filename)
        if not label or label == "normal":
            continue
        if not include_hidden and filename in hidden:
            continue
        stat = await asyncio.to_thread((RECORDINGS_DIR / filename).stat)
        items.append({
            "filename": filename,
            "label": label,
            "createdAt": to_iso_millis(stat.st_mtime),
            "size": stat.st_size,
            "hidden": filename in hidden,
        })

    items.sort(key=lambda item: item["createdAt"], reverse=True)
    return items
